sql: fix CREATE TABLE LIKE with implicit pk

jasonmchan opened the pull request:
Previously, `CREATE TABLE LIKE` copied implicitly created columns (e.g.
for the rowid default primary key and hash sharded index). Defaults for
some of these columns were not properly copied over in some cases,
causing unexpected constraint violations to surface.

This commit fixes this by skipping copying such columns; instead, they
will be freshly created. Followup work is needed for REGIONAL BY ROW.

Fixes #82401

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82555)
<!-- Reviewable:end -->

chengxiong-ruan reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82555#-:-N4-HxLk0b2kPe-yHmAp:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

otan commented:
probably worth considering https://github.com/cockroachdb/cockroach/issues/82401#issuecomment-1149675011 before merging this

knz commented:
That solution is not sufficient.  See my additional comment in https://github.com/cockroachdb/cockroach/issues/82401.

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82555)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @jasonmchan)

___
*[`pkg/sql/create_table.go` line 2894 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82555#-N45Nx1dEtxLCnMd5V3m:-N45Nx1dEtxLCnMd5V3n:bm0d82s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cd587fc9e25ec975ff54bf1301d86862685ba0e/pkg/sql/create_table.go#L2894)):*
> ```Go
> 		return false, err
> 	}
> 	if td.IsShardColumn(col) {
> ```

Could you also add logic tests where the source table has `hash sharded index` and `expression index`?


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/create_table.go` line 2894 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82555#-N45Nx1dEtxLCnMd5V3m:-N45P9YyCMTh6XJxHbVe:b-5i04t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cd587fc9e25ec975ff54bf1301d86862685ba0e/pkg/sql/create_table.go#L2894)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Could you also add logic tests where the source table has `hash sharded index` and `expression index`?
</blockquote></details>

oops, there is already test coverage for `hash sharded index`.


<!-- Sent from Reviewable.io -->

jasonmchan commented:
I'm going to omit a fix for REGIONAL BY ROW because #82672 causes problems with this solution. Will put up a new revision for the broken CI tests.

Commit:
sql: fix CREATE TABLE LIKE with implicit pk

Previously, `CREATE TABLE LIKE` copied implicitly created columns (e.g.
for the rowid default primary key and hash sharded index). Defaults for
some of these columns were not properly copied over in some cases,
causing unexpected constraint violations to surface.

This commit fixes this by skipping copying such columns; instead, they
will be freshly created. Followup work is needed for REGIONAL BY ROW.

Fixes #82401

Release note: None

chengxiong-ruan reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
I'd suggest adding a test case for expression index like `INDEX idx_expr (a + 2)` which would create a hidden virtual column which can be seen by selecting from the `crdb_internal.table_columns` table. I think this is working with PR but up to you to add it or not :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82555#-:-N48e4Tt0KPuxbHbqexU:bhfjljv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @jasonmchan)



<!-- Sent from Reviewable.io -->

jasonmchan commented:
TFTR! I'm leaving out the expression index test for now because this PR doesn't change how they are handled.

bors r+
